src/Runtime/OMTensor.inc

```diff
@@ -0,0 +1,434 @@
+//===--------- OMTensor.inc - C/C++ Neutral OMTensor Implementation--------===//
+//
+// Copyright 2019-2020 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file contains implementations of OMTensor data structures
+// and helper functions.
+//
+//===----------------------------------------------------------------------===//
+
+#ifdef __cplusplus
+#include <cassert>
+#include <map>
+#include <numeric>
+#include <random>
+#include <string>
+#include <typeinfo>
+#include <vector>
+#else
+#include <assert.h>
+#endif
+
+#ifdef __APPLE__
+#include <stdlib.h>
+#else
+#include <malloc.h>
+#endif
+
+#include <stdio.h>
+#include <string.h>
+
+#include "onnx-mlir/Runtime/OMTensor.h"
+
+struct OMTensor {
+#ifdef __cplusplus
+  /**
+   * Constructor
+   *
+   * @param rank, rank of data sizes and strides
+   *
+   * Create a OMTensor with specified rank. Memory for data sizes and strides
+   * are allocated.
+   */
+  OMTensor(int rank) {
+    if ((_shape = (int64_t *)malloc(rank * sizeof(int64_t))) &&
+        (_stride = (int64_t *)malloc(rank * sizeof(int64_t)))) {
+      _allocatedPtr = NULL;
+      _alignedPtr = NULL;
+      _offset = 0;
+      _dataType = ONNX_TYPE_UNDEFINED;
+      _rank = rank;
+      _owning = false;
+    } else {
+      throw std::runtime_error(
+          "OMTensor(" + std::to_string(rank) + ") malloc error");
+    }
+  };
+
+  OMTensor() = default;
+
+  /**
+   * Destructor
+   *
+   * Destroy the OMTensor struct.
+   */
+  ~OMTensor() {
+    if (_owning)
+      free(_allocatedPtr);
+    free(_shape);
+    free(_stride);
+  };
+#endif
+  // Fields are named according to:
+  // https://mlir.llvm.org/docs/Dialects/SPIR-V/#lowering-memrefs-to-spvarray-and-spvrtarray
+
+  void *_allocatedPtr; // data buffer
+  void *_alignedPtr;   // aligned data buffer that the omt indexes.
+
+  int64_t _offset;  // offset of 1st element
+  int64_t *_shape;  // sizes array
+  int64_t *_stride; // strides array
+  int64_t _rank;    // rank
+
+  OM_DATA_TYPE _dataType; // ONNX data type
+
+  int _owning; // indicates whether the Omt owns the memory space
+               // referenced by _allocatedPtr. Omt struct will release the
+               // memory space referred to by _allocatedPtr upon destruction if
+               // and only if it owns it.
+};
+
+// Create a OMTensor.
+OMTensor *omTensorCreate(
+    void *data_ptr, int64_t *shape, int64_t rank, OM_DATA_TYPE dtype) {
+  OMTensor *tensor = (OMTensor *)malloc(sizeof(OMTensor));
+  if ((tensor->_shape = (int64_t *)malloc(rank * sizeof(int64_t))) &&
+      (tensor->_stride = (int64_t *)malloc(rank * sizeof(int64_t)))) {
+    tensor->_allocatedPtr = data_ptr;
```

Comment:
Need to check each `malloc` for failure. And if one fails, the previous successful `malloc` must be freed.

Reply:
If we bail out entirely (i.e., assertion failure with no recovery), it seems like we can get away with letting the OS free up the memory we malloc'ed? Otherwise the code gets complicated since we have three malloc's in a row here.

Reply:
nvm, saw your comments on the return value, we actually don't bail out.

Reply:
Yes assertions are for catching bugs, i.e., internal code logic error. Malloc failures are not bugs. And APIs typically don't end abruptly (unless it's an assertion failure due to a bug of course) since they should give the callers a chance to clean up.

BTW, the logic for checking the malloc failures should be fairly simple. It can be some thing like:
```
if (tensor == NULL) return NULL;
if ((tensor->_shape = (int64_t *)malloc(rank * sizeof(int64_t))) == NULL) {
  free(tensor);
  return NULL;
}
if ((tensor->_stride = (int64_t *)malloc(rank * sizeof(int64_t))) == NULL) {
  free(tensor->_shape);
  free(tensor);
  return NULL;
}
...
```